components/codec/number.rs

```diff
@@ -0,0 +1,2096 @@
+                            &cursor.get_ref().as_slice()[0..pos],
+                            &payload.as_slice()[0..pos]
+                        );
+                        assert_eq!(
+                            &cursor.get_ref().as_slice()[pos..pos + len],
+                            &base_buf.as_slice()[0..len]
+                        );
+                        assert_eq!(
+                            &cursor.get_ref().as_slice()[pos + len..],
+                            &payload.as_slice()[pos + len..]
+                        );
+                        assert_eq!(cursor.position(), (pos + len) as u64);
+                    }
+                }
+            }
+        };
+    }
+
+    #[test]
+    fn test_var_u64() {
+        test_varint!(
+            get_u64_samples(),
+            encode_var_u64,
+            try_decode_var_u64,
+            write_var_u64,
+            read_var_u64,
+        );
+        // Check encoded output with ProtoBuf
+        let samples = get_u64_samples();
+        for sample in samples {
+            let mut buf = vec![0; super::MAX_VARINT64_LENGTH];
+            let len = super::NumberCodec::encode_var_u64(buf.as_mut_slice(), sample);
+
+            let mut protobuf_output = vec![];
+            {
+                let mut writer = CodedOutputStream::new(&mut protobuf_output);
+                writer.write_uint64_no_tag(sample).unwrap();
+                writer.flush().unwrap();
+            }
+            assert_eq!(buf[..len], protobuf_output[..]);
+        }
+    }
+
+    #[test]
+    fn test_var_i64() {
+        test_varint!(
+            get_i64_samples(),
+            encode_var_i64,
+            try_decode_var_i64,
+            write_var_i64,
+            read_var_i64,
+        );
+    }
+}
+
+#[cfg(test)]
+mod benches {
+    use test;
+
+    use byteorder;
+    use protobuf::CodedOutputStream;
+
+    /// Encode u64 little endian using `NumberCodec` and store position in extra variable.
+    #[bench]
+    fn bench_encode_u64_le_number_codec(b: &mut test::Bencher) {
+        let mut buf: [u8; 10] = [0; 10];
+        b.iter(|| {
+            let mut pos = 0;
+            super::NumberCodec::encode_u64_le(
+                test::black_box(&mut buf[..]),
+                test::black_box(0xDEADBEEF),
+            );
+            pos += 8;
+            test::black_box(buf);
+            assert_eq!(pos, 8);
+        });
+    }
+
+    /// Encode u64 little endian using `byteorder::WriteBytesExt` over a `Cursor<&mut [u8]>`.
+    #[bench]
+    fn bench_encode_u64_le_byteorder(b: &mut test::Bencher) {
+        use byteorder::WriteBytesExt;
+
+        let mut buf: [u8; 10] = [0; 10];
+        b.iter(|| {
+            {
+                let mut cursor = ::std::io::Cursor::new(test::black_box(&mut buf[..]));
+                cursor
+                    .write_u64::<byteorder::LittleEndian>(test::black_box(0xDEADBEEF))
+                    .unwrap();
+                assert_eq!(cursor.position(), 8);
```

Comment:
You should avoid use `assert` in benchmark body, which can affect the performance.